File: src/main/java/kr/co/bookand/backend/config/security/PrincipalDetailService.java
package kr.co.bookand.backend.config.security;

import kr.co.bookand.backend.account.domain.Account;
import kr.co.bookand.backend.account.exception.AccountException;
import kr.co.bookand.backend.account.repository.AccountRepository;
import kr.co.bookand.backend.common.exception.ErrorCode;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class PrincipalDetailService implements UserDetailsService {

    private final AccountRepository accountRepository;

    @Override
    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
        return accountRepository.findByEmailAndVisibilityTrue(email)
                .map(this::createUserDetails)
                .orElseThrow(() -> new AccountException(ErrorCode.NOT_FOUND_MEMBER, email));
    }
    private PrincipalDetails createUserDetails(Account account){
        return new PrincipalDetails(account);
    }

}

File: src/main/kotlin/kr/co/bookand/backend/config/security/PrincipalDetailService.kt
package kr.co.bookand.backend.config.security

import kr.co.bookand.backend.account.domain.Account
import kr.co.bookand.backend.account.repository.AccountRepository
import kr.co.bookand.backend.common.ErrorCode
import kr.co.bookand.backend.common.exception.BookandException
import lombok.RequiredArgsConstructor
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.stereotype.Service

@Service
@RequiredArgsConstructor
class PrincipalDetailService(
    private val accountRepository: AccountRepository
) : UserDetailsService {
    override fun loadUserByUsername(email: String): UserDetails {
        val account = accountRepository.findByEmail(email)
            ?: throw BookandException(ErrorCode.NOT_FOUND_MEMBER)
        return createUserDetails(account)
    }

    private fun createUserDetails(account: Account): UserDetails {
        return PrincipalDetails(account)
    }

}